### PR TITLE
Set aperture only if value changes

### PR DESCRIPTION
**Pull request by ebadkamil**
Fixes the bug reported [here](https://github.com/ess-dmsc/nicos/pull/48#issuecomment-768967296).

### CODE REVIEW

**Comment by ebadkamil** [comment]
I would prefer a `SmartLineEdit` that changes color when user is editing. This will force them to `press Enter` to register. This will require some work and can be done in another PR.   `editingFinished` signal is good but fired too often like tabbing, clicking away etc.

File: nicos_ess/loki/gui/sampleconf.py [hunk 1]
```diff
@@ -502,44 +500,66 @@ def on_list_itemClicked(self, item):
...
+            return
+        self._set_apperture_value_at_key(index, 2, value)
+    
+    def set_height_h(self, index, value):
+        value = float(value)
+        if self.configs[index]['aperture'][3] == value:
+            return
+        self._set_apperture_value_at_key(index, 3, value)
+    
+    def _set_apperture_value_at_key(self, index, key, value):
```
Comment (mattclarke) [inline nicos_ess/loki/gui/sampleconf.py:557]: Typo, one p in aperture

Comment (ebadkamil) [reply nicos_ess/loki/gui/sampleconf.py:557]: yeps. thanks

File: nicos_ess/loki/gui/sampleconf.py [hunk 2]
```diff
@@ -614,13 +633,14 @@ def _save_script(filename, script):
...
+    # May be a better name.
+    def _set_dirty(self, value: bool):
         """
         Enables apply changes button if there are changes in the sample
         configuration.
         """
-        self.dirty = True
-        self.applyBtn.setEnabled(True)
+        self.dirty = value
+        self.applyBtn.setEnabled(value)
```
Comment (mattclarke) [inline nicos_ess/loki/gui/sampleconf.py:643]: If self.dirty and self.applyBtn.enabled always change together then perhaps we don't need self.dirty?

**Comment by mattclarke** [comment]
The apply button should become active the moment I type in one of the aperture boxes. Currently, I have to click away from it instead.
Probably makes sense to create a separate ticket for that.

**Comment by ebadkamil** [comment]
Agree.

File: nicos_ess/loki/gui/sampleconf.py (hunk 2 above)
Comment (ebadkamil) [reply nicos_ess/loki/gui/sampleconf.py:643]: I think it doesn't. At few place `self.dirty` is only set.

Comment (mattclarke) [reply nicos_ess/loki/gui/sampleconf.py:643]: That sounds like a legacy code smell.
It would be better if we only had one way of doing it.

Comment (ebadkamil) [reply nicos_ess/loki/gui/sampleconf.py:643]: Ok let me check.

Comment (jkrueger1) [reply nicos_ess/loki/gui/sampleconf.py:643]: I have the feeling that 'dirty' is simply used to mark, that the self.applyBtn is enabled or not. But if the button isn't enabled the user can't perform some action. So it's useful to verify in what cases the user should be allowed to apply or not.

Comment (ebadkamil) [reply nicos_ess/loki/gui/sampleconf.py]: `dirty` was set when one `retrieves_current` or `import_from_files`. and was used as a validator to submit the script generated. It was also used when toggling between view mode. Now I have removed `dirty` and made simple changes to get the desired behavior. Please check if it makes sense.

File: nicos_ess/loki/gui/sampleconf.py [hunk 3]
```diff
@@ -256,11 +256,10 @@ def __init__(self, parent, client, options):
         self.createBtn.setMenu(menu)
 
         self.configs = []
-        self.dirty = False
-        self.applyBtn.setEnabled(False)
         self.holder_info = options.get('holder_info', [])
         self.instrument = options.get('instrument', 'loki')
-
+        self.applyBtn_was_enabled = False
```
Comment (mattclarke) [inline nicos_ess/loki/gui/sampleconf.py:261]: `unapplied_changes` would be a better name, even though unapplied isn't a word.

Comment (mattclarke) [reply nicos_ess/loki/gui/sampleconf.py:261]: At least, according to FireFox's spell-checker

File: nicos_ess/loki/gui/sampleconf.py [hunk 4]
```diff
@@ -502,44 +501,66 @@ def on_list_itemClicked(self, item):
...
+            return
+        self._set_aperture_value_at_key(index, 0, value)
+        
+    def set_pos_y(self, index, value):
+        value = float(value)
+        if self.configs[index]['aperture'][1] == value:
+            return
+        self._set_aperture_value_at_key(index, 1, value)
+    
+    def set_width_w(self, index, value):
```
Comment (mattclarke) [inline nicos_ess/loki/gui/sampleconf.py:546]:
```suggestion
    def set_width(self, index, value):
```
?

File: nicos_ess/loki/gui/sampleconf.py [hunk 5]
```diff
@@ -502,44 +501,66 @@ def on_list_itemClicked(self, item):
...
+        value = float(value)
+        if self.configs[index]['aperture'][2] == value:
+            return
+        self._set_aperture_value_at_key(index, 2, value)
+    
+    def set_height_h(self, index, value):
```
Comment (mattclarke) [inline nicos_ess/loki/gui/sampleconf.py:552]: Can we lose the _h?

File: nicos_ess/loki/gui/sampleconf.py (hunk 4 above)
Comment (mattclarke) [inline nicos_ess/loki/gui/sampleconf.py:546]: Can we lose the `-w`?

File: nicos_ess/loki/gui/sampleconf.py [hunk 6]
```diff
@@ -502,44 +501,66 @@ def on_list_itemClicked(self, item):
         layout = self.frame.layout()
         layout.addWidget(frm)
 
+        # editingFinished signal is fired when LineEdit lose focus.
```
Comment (ebadkamil) [inline nicos_ess/loki/gui/sampleconf.py:504]: We can get rid of these comments once the review is complete.

**Comment by mattclarke** [comment]
Okay, I think this is ready. @ebadkamil have one last look through to make sure you are happy and once the build finishes feel free to merge.

**Comment by ebadkamil** [comment]
Thanks. Works and looks all good.

**Pull request merged**